staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -300,3 +307,61 @@ func (le *LeaderElector) maybeReportTransition() {
 		go le.config.Callbacks.OnNewLeader(le.reportedLeader)
 	}
 }
+
+type LeaderChecker interface {
+	healthz.HealthzChecker
+	SetLeaderElection(le *LeaderElector)
+}
+
+type timeNow func() time.Time
+
+type leaderChecker struct {
+	c           *apiserverconfig.LeaderElectionConfiguration
+	pointerLock sync.Mutex
+	le          *LeaderElector
+	timeNow
+}
+
+func (l *leaderChecker) Name() string {
+	return "leaderElection/" + l.c.ResourceLock
+}
+
+func (l *leaderChecker) Check(req *http.Request) error {
+	if !l.c.LeaderElect {
+		return nil
+	}
+	l.pointerLock.Lock()
+	defer l.pointerLock.Unlock()
+	if l.le == nil {
+		return nil
+	}
+	if !l.le.IsLeader() {
+		// Currently not concerned with the case that we are hot standby
+		return nil
+	}
+	now := l.timeNow()
+	// If we are more than 20 seconds after the lease durection that is past the timeout on the lease renew.
+	// Time to start reporting ourselves as unhealthy. We should have died but conditions like deadlock can
+	// prevent this. (See #70819)
+	expiry := l.le.observedTime.Add(l.le.config.LeaseDuration).Add(time.Second * 20)
```

Comment:
If you add a function that exposes observedTime (or the observedRecord's acquireTime), then maybe you can put all of this code in another package?

Reply:
I could but I don't want to expose the inner workings of leaderelection as I feel it should remain obfuscated. The more we expose these details the less room we have to fix it later. I also think that having the optional healthz check which is responsible for/understands the inner workings of leaderelection is a good thing. I agree with not wanting to pull in server code to the client package but I think you other suggestions have essentially accomplished that.